### PR TITLE
Context-sensitive auto completion

### PR DESCRIPTION
**Pull request by asuhan**
Bring back the Calcite-based completion with improvements. It now
follows the permissions, can auto-complete grouped columns and doesn't
conflate column and table suggestions anymore.

### CODE REVIEW

**Comment by micahstubbs** [comment]
@asuhan very nice, cool to see the progress.

curious, is there any change in the `get_completion_hints` thrift endpoint in this PR? (would I need to update my client code to test this out)?

**Comment by asuhan** [comment]
No, the interface is exactly the same.

File: ThriftHandler/MapDHandler.cpp [hunk 1]
```diff
@@ -720,35 +720,79 @@ void MapDHandler::get_completion_hints(std::vector<TCompletionHint>& hints,
                                        const TSessionId& session,
                                        const std::string& sql,
                                        const int cursor) {
-  const auto last_word = find_last_word_from_cursor(sql, cursor < 0 ? sql.size() : cursor);
+  get_completion_hints_unsorted(hints, session, sql, cursor);
+  // Sort the hints by category, from COLUMN (most specific) to KEYWORD.
+  std::sort(hints.begin(), hints.end(), [](const TCompletionHint& lhs, const TCompletionHint& rhs) {
+    return lhs.type < rhs.type;
+  });
+}
```
Comment (micahstubbs) [inline ThriftHandler/MapDHandler.cpp]: oh, I like this hint sort order.  this is something that @tmostak and I had also talked about 👌

**Comment by andrewseidl** [comment]
@asuhan did we also want to pull in https://github.com/mapd/mapd-core/commit/653ae6775b94fd9bd6c5606cb2666a7d1547e62f from branch `alex/mapdql_completions` here?

**Comment by asuhan** [comment]
@andrewseidl Up to you. It's convenient and should be ok, but I haven't tested it extensively.

**Comment by andrewseidl** [comment]
Holding off on `alex/mapdql_completions` until #147 is merged, I'll take care of resolving the conflicts once it's in.

**Pull request merged**